WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift

```diff
@@ -19,30 +19,177 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+                                                                          pageNumber: 1,
+                                                                          pageSize: Constants.numberOfItems) { result in
+                continuation.resume(with: result)
+            })
+        }
+    }
+
+    @MainActor
+    func retrieveProducts(for reviewProductIDs: [Int64]) async throws -> (products: [Product], hasNextPage: Bool) {
+        try await withCheckedThrowingContinuation { continuation in
+            stores.dispatch(ProductAction.retrieveProducts(siteID: siteID,
+                                                           productIDs: reviewProductIDs
+                                                          ) { result in
+                continuation.resume(with: result)
+            })
+        }
+    }
+
+    /// Performs initial fetch from storage and updates results.
+    func configureResultsControllers() {
+        configureProductReviewsResultsController()
+        configureProductsResultsController()
+    }
+
+    func configureProductReviewsResultsController() {
+        productReviewsResultsController.onDidChangeContent = { [weak self] in
+            guard let self else { return }
+            Task {
+                await self.updateReviewsResults()
+            }
+        }
+        productReviewsResultsController.onDidResetContent = { [weak self] in
+            guard let self else { return }
+            Task {
+                await self.updateReviewsResults()
+            }
+        }
+
+        do {
+            try productReviewsResultsController.performFetch()
+        } catch {
+            ServiceLocator.crashLogging.logError(error)
+        }
+    }
+
+    func configureProductsResultsController() {
+        productsResultsController.onDidChangeContent = { [weak self] in
+            self?.updateProductsResults()
+        }
+        productsResultsController.onDidResetContent = { [weak self] in
+            self?.updateProductsResults()
+        }
+
+        do {
+            try productsResultsController.performFetch()
+        } catch {
+            ServiceLocator.crashLogging.logError(error)
+        }
+    }
+
+    /// Updates data
+    @MainActor
+    func updateReviewsResults() async {
+        let reviews = productReviewsResultsController.fetchedObjects.prefix(Constants.numberOfItems)
+        let reviewProductIDs = reviews.map { $0.productID }
+
+        // Load products that matches the review product IDs
+        if reviewProductIDs.isNotEmpty {
+            await loadReviewProducts(for: reviewProductIDs)
+        }
+
+        data = reviews
+            .map { review in
+                let product = reviewProducts.first { $0.productID == review.productID }
+                // TODO: also fetch notification
+                return ReviewViewModel(review: review, product: product, notification: nil)
+            }
+    }
```

Comment:
I'd like to get more opinion on this area.

The fetch result for reviews only includes data for product ID related to the reviews, while the UI requires the product name to be displayed. Thus, another call for loadReviewProducts() is needed to get the product name.

I'm not 100% sure if this is the best way to do this. It feels to me that it's going through many steps, but seems inevitable if we want to use storage as single source of truth:

1. remote call for reviews
2. handle db update for reviews
    1. remote call for products
    2. handle db update for products
    3. compare local reviews and local products to build data

Reply:
Thanks for explaining, Hafiz!

I think you can try fetching from storage before sending a remote call for products.

You can use the following predicate to fetch multiple products from storage.

```
NSPredicate(format: "siteID == %lld AND productID IN %@", siteID, arrayOfProductIDs)
```

What do you think?

Reply:
I updated the logic in https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/4ad2e89c4dcf4bd4c51f0ba8662446e96aad4ccd 

Now it will fetch products from storage first and only do remote call if they're not found.